### DISCUSSION TITLE
WebUSB for firmware upgrade

### DISCUSSION THREAD

**Discussion by bademux** [joeycastillo/Sensor-Watch#129] [General]
Hi,
I would like to propose WebUSB for firmware upgrade (something like https://devanlai.github.io/webdfu/dfu-util/).
We already have emscripten simulator and i will be convenient to upload firmware from the page.

In my opinion current approach with emulation is not optimal: 
1. It can't be automated with dedicated service for building and generating firmware. User have to copy firmware manually.
2. Not cross platform, a lot of effort need to be invest in tooling. Example  `make install` has assumption that are not correct, like all usb drives are mounted in `/media`

I found @joeycastillo mentioned WebUSB in context of configuration https://github.com/joeycastillo/Sensor-Watch/pull/23#issuecomment-976737718

In my opinion current cmd based approach is more elastic\can be adapted to wireless communication channel.
Unfortunately there is  a bug https://github.com/joeycastillo/Sensor-Watch/issues/117, but it is fixable I think.

**Comment by wryun** [comment]
Not sure exactly what you're looking for, but perhaps https://sensor-watch-builder.fly.dev/ is interesting to you. One day I will clean it up :)

It's very simple: https://github.com/wryun/sensor-watch-builder/

I don't think that it's a big bar to drag and drop the firmware onto the drive; the user has already had to pull apart their watch!

**Comment by bademux** [comment]
"Not sure exactly what you're looking for, but perhaps https://sensor-watch-builder.fly.dev/ is interesting to you. One day I will clean it up :)"
@wryun I had the same idea, and then saw 2 separate impls on discord :)


"I don't think that it's a big bar to drag and drop the firmware onto the drive"

Why to not use all goods of new technology quite some. I'm not sure about using "Then download firmware and copy it into fake flash drive", instead of just providing the button "Flash". Besides fake usb drive will be constant source of questions like "Can I copy my documents onto it?".
It really convenient to flash from web browser especially with simulator. Firmware can be tested and flashed instantly.

My second point is that automation for `make install` is straight gate to hell due to multiple OS support (having fun with it :( ).

**Comment by tahnok** [comment]
It could be interesting to look into but

https://developer.mozilla.org/en-US/docs/Web/API/USB

Shows that Firefox does not support WebUSB

**Reply by bademux** [reply]
@tahnok don't now what to say to not change subject to Chromium (all Bilink based browsers) vs the rest.
My point was that there is software that can provide us OS abstraction layer for USB, named  Chromium.
sidenote:  I'm a firefox user.

Thanks for advance for the replay. 
Looks like current deploy through USB drive will stay with us, so I will try to find time to propose update hard coded `/media` for some semi-popular linuxes like Arch.
